File: csharp/pythagorean-triplet/PythagoreanTriplet.cs
using System;
using System.Collections.Generic;

public class Triplet
{
  private int a;
  private int b;
  private int c;

  public Triplet(int a, int b, int c)
  {
    this.a = a;
    this.b = b;
    this.c = c;
  }

  public int Sum()
  {
    return a + b + c;
  }

  public int Product()
  {
    return a * b * c;
  }

  public bool IsPythagorean()
  {
    return Math.Pow(a, 2) + Math.Pow(b, 2) == Math.Pow(c, 2);
  }

  public static IEnumerable<Triplet> Where(int maxFactor, int minFactor = 1, int sum = 0)
  {
    var triplets = new List<Triplet>();
    for (int i = minFactor; i < maxFactor - 1; i++)
    {
      for (int j = i + 1; j < maxFactor; j++)
      {
        for (int k = j + 1; k <= maxFactor; k++)
        {
          Triplet triplet = new Triplet(i, j, k);
          if (ValidTriplet(triplet, sum))
          {
            triplets.Add(triplet);
          }
        }
      }
    }
    return triplets;
  }

  private static bool ValidTriplet(Triplet triplet, int sum)
  {
    return triplet.IsPythagorean() && (triplet.Sum() == sum || sum == 0);
  }
}

File: javascript/pythagorean-triplet/pythagorean-triplet.js
export const triplets = ({ minFactor, maxFactor, sum }) => {
  minFactor = minFactor ?? 1;
  maxFactor = maxFactor ?? sum - minFactor;
  let trips = [];
  for (let i = minFactor; i < maxFactor; i++) {
    for (let j = i + 1; j < maxFactor; j++) {
      let k = sum - i - j;
      if (k <= maxFactor && i ** 2 + j ** 2 === k ** 2)
        trips.push(new Triplet(i, j, k));
    }
  }
  return trips;
}

class Triplet {
  constructor(a, b, c) {
    this._a = a;
    this._b = b;
    this._c = c;
  }

  toArray = () => [this._a, this._b, this._c];
}
